README.md
<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
**Table of Contents**

- [@niteshp/ngx-stripe](#niteshpurohitngx-stripe)
  - [Features](#features)
  - [Installation](#installation)
  - [Using the library](#using-the-library)
  - [StripeCardComponent](#stripecardcomponent)
    - [RequestPaymentButton](#requestpaymentbutton)
    - [Dynamic Stripe Keys](#dynamic-stripe-keys)
    - [PaymentIntents](#paymentintents)
    - [SCA/3D Secure payments](#sca3d-secure-payments)
  - [Testing](#testing)
  - [Building](#building)
  - [Publishing](#publishing)
  - [Documentation](#documentation)
  - [License](#license)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

# @niteshp/ngx-stripe

[![version](https://img.shields.io/npm/v/ngx-stripe.svg)](https://www.npmjs.com/package/@niteshp/ngx-stripe)
[![license](https://img.shields.io/npm/l/express.svg)](https://www.npmjs.com/package/@niteshp/ngx-stripe)

@angular wrapper for StripeJS based on module by Ricardo Sánchez Gregorio at https://github.com/richnologies/ngx-stripe

_Note: This version is designed to work with angular v8+ only_

## Features

* Stripe Service
* Lazy script loading
* AoT compliant
* SSR fallback

## Installation

To install this library, run:

```bash
$ npm install @niteshp/ngx-stripe --save
or
$ yarn add @niteshp/ngx-stripe
```

## Using the library

Import the `NgxStripeModule` into the application

The module takes the same parameters as the global Stripe object. The APIKey and the optional options to use Stripe connect

* apiKey: string
* options: {
  stripeAccount?: string;
}

```typescript
import { BrowserModule } from '@angular/platform-browser';
import { NgModule } from '@angular/core';

import { AppComponent } from './app.component';

// Import your library
import { NgxStripeModule } from '@niteshp/ngx-stripe';

@NgModule({
  declarations: [
    AppComponent
  ],
  imports: [
    BrowserModule,
    NgxStripeModule.forRoot('***your-stripe-publishable key***'),
    LibraryModule
  ],
  providers: [],
  bootstrap: [AppComponent]
})
export class AppModule { }
```

_Note: If you leave the publishable key blank you must set on using StripeSerive.changeKey prior to creating token's or sources_

## StripeCardComponent

As an alternative to the previous example, you could use the StripeCardComponent.

It will make a little bit easier to mount the card.

To fetch the Stripe Element, you could you use either the (change) output, or,
by using a ViewChild, the public method getCard()

//stripe.compnent.html
```html
<ngx-stripe-card [options]="cardOptions" [elementsOptions]="elementsOptions" (change)="cardUpdated($event)" (error)="error = $event"></ngx-stripe-card>
<div class="error">
  {{error?.message}}
</div>
<button (click)="getCardToken()" [disabled]="!complete">Get Card Token</button>
```

//stripe.component.ts
```typescript
import { Component, OnInit, ViewChild, ElementRef } from '@angular/core';
import { FormGroup, FormBuilder, Validators } from "@angular/forms";

import { StripeService, StripeCardComponent, ElementOptions, ElementsOptions, Element as StripeElement  } from "@niteshp/ngx-stripe";

@Component({
  selector: 'app-stripe-test',
  templateUrl: 'stripe.component.html'
})
export class StripeTestComponent implements OnInit {
  stripeKey = '';
  error: any;
  complete = false;
  element: StripeElement;
  cardOptions: ElementOptions = {
    style: {
      base: {
        iconColor: '#276fd3',
        color: '#31325F',
        lineHeight: '40px',
        fontWeight: 300,
        fontFamily: '"Helvetica Neue", Helvetica, sans-serif',
        fontSize: '18px',
        '::placeholder': {
          color: '#CFD7E0'
        }
      }
    }
  };

  elementsOptions: ElementsOptions = {
    locale: 'en'
  };

  constructor(
    private _stripe: StripeService
  ) {}

  cardUpdated(result) {
    this.element = result.element;
    this.complete = result.card.complete;
    this.error = undefined;
  }

  keyUpdated() {
    this._stripe.changeKey(this.stripeKey);
  }

  getCardToken() {
    this._stripe.createToken(this.element, {
      name: 'tested_ca',
      address_line1: '123 A Place',
      address_line2: 'Suite 100',
      address_city: 'Irving',
      address_state: 'BC',
      address_zip: 'VOE 1H0',
      address_country: 'CA'
    }).subscribe(result => {
      // Pass token to service for purchase.
      console.log(result);
    });
  }
}
```

### RequestPaymentButton
A payment request button (e.g. apple pay, google chrome payment, etc.) can be enabled with the `ngx-payment-request` component. All of the options [Request Payment Api](https://stripe.com/docs/payment-request-api) options are exposed via the three options sections in the component.

// request-button.component.html
```html
<div class="container">
  <mat-card>
    Pay: <mat-slider min="1" max="100" step="10" thumbLabel [value]="pay" (valueChange)="updatePay($event)"></mat-slider>
    <ngx-payment-request
      [options]="requestOptions"
      [elementsOptions]="elementsOptions"
      [styles]="styles"
      (shippingAddressChange)="updateShippingAddress($event)"
      (shippingOptionChange)="updateShippingOption($event)"
      (change)="requestUpdated($event)">
    </ngx-payment-request>

  </mat-card>
```

// request-button.compnent.ts
```typescript
@Component({
  selector: 'app-test',
  templateUrl: 'request-button.component.html'
})
export class RequestButtonComponent implements OnInit {
  @Input set stripeKey(key) {
    this._stripe.changeKey(key);
  }

  public requestOptions: RequestElementOptions = {
    country: 'US',
    currency: 'usd',
    requestPayerName: true,
    requestPayerEmail: true,
    requestPayerPhone: true,
    requestShipping: true,
    total: {
      amount: 10000,
      label: 'Donate to the things'
    }
  };

  public styles: PaymentRequestButtonStyle = {
    type: 'donate',
    theme: 'light',
    height: '64px'
  };

  public elementsOptions: ElementsOptions = {
    locale: 'en'
  };

  public pay = 100;

  constructor(private _stripe: StripeService) { }

  public updatePay(amount: number) {
    this.pay = amount;
    this.requestOptions = {
      ...this.requestOptions,
      total: {
        ...this.requestOptions.total,
        amount: amount * 100
      }
    };
  }

  public async requestUpdated(result) {
    const response = await fetch('/charges', {
      method: 'POST',
      body: JSON.stringify({token: result.token.id}),
      headers: {'content-type': 'application/json'},
    });

    if (response.ok) {
      result.complete('success');
    } else {
      result.complete('fail');
    }
  }

  public updateShippingAddress(result) {
    result.updateWith({
      status: 'success',
      shippingOptions: [
        {
          id: 'free-shipping',
          label: 'Free shipping',
          detail: 'Arrives in 5 to 7 days',
          amount: 0
        },
        {
          id: 'express',
          label: 'Express shipping',
          detail: 'Arrives in 1 to 2 days',
          amount: 1000
        }
      ]
    });
  }

  public updateShippingOption(result) {
    result.updateWith({
      status: 'success'
    });
  }
}

```

### Dynamic Stripe Keys
Given the way an AOT angular build processes provided values if you need to provide a stripe key at runtime you'll want to forgo the forRoot initialization and instead add a factory and provider

```ts

  export function NgxStripeFactory(): string {
    return dynamicLocation.stripePkKey;
  }

  @NgModule({
    ...
    imports: []
      ...
      NgxStripeModule.forRoot('whatever key you think should be default'),
      ...
    ]
    providers: [
      {
        provide: STRIPE_PUBLISHABLE_KEY,
        useFactory: NgxStripeFactory
      }
    ]
  })
```

### PaymentIntents
You can use the payment intents api with this service just like with stripe.js the following are exposed via the `StripeService`

```ts
  handleCardSetup(clientSecret: string, el: Element, cardSetupOptions?: SetupIntentData): Promise<SetupIntentResult>;
  handleCardAction(clientSecret: string);
  handleCardPayment(clientSecret: string, el: Element, data: CardPaymentData);
  confirmPaymentIntent(clientSecret: string, el: Element, data: ConfirmIntentData);
  retrievePaymentIntent(clientSecret: string);
  confirmSetupIntent(clientSecret: string, el: Element, data: ConfirmSetupIntentData);
  retrieveSetupIntent(clientSecret: string);
```

### SCA/3D Secure payments

3D Secure payments are enabled by following a [3D Secure Flow](https://stripe.com/docs/payments/3d-secure#example-of-a-3d-secure-2-flow) using the payment intents api's. You will need to setup a proper next action in the payment intent based off of the 3D Secure response from the customers bank/institution.

Example:
//stripe.compnent.html
```html
<ngx-stripe-card [options]="cardOptions" [elementsOptions]="elementsOptions" (change)="cardUpdated($event)" (error)="error = $event"></ngx-stripe-card>
<div class="error">
  {{error?.message}}
</div>
<button (click)="payNow()" [disabled]="!complete">Pay</button>
```

//stripe.component.ts
```typescript
import { Component, OnInit, ViewChild, ElementRef } from '@angular/core';
import { FormGroup, FormBuilder, Validators } from "@angular/forms";

import { StripeService, StripeCardComponent, ElementOptions, ElementsOptions, Element as StripeElement } from "@niteshp/ngx-stripe";

@Component({
  selector: 'app-stripe-test',
  templateUrl: 'stripe.component.html'
})
export class StripeTestComponent implements OnInit {
  stripeKey = '';
  error: any;
  complete = false;
  element: StripeElement;
  cardOptions: ElementOptions = {
    style: {
      base: {
        iconColor: '#276fd3',
        color: '#31325F',
        lineHeight: '40px',
        fontWeight: 300,
        fontFamily: '"Helvetica Neue", Helvetica, sans-serif',
        fontSize: '18px',
        '::placeholder': {
          color: '#CFD7E0'
        }
      }
    }
  };

  elementsOptions: ElementsOptions = {
    locale: 'en'
  };

  constructor(
    private _stripe: StripeService
  ) {}

  cardUpdated(result) {
    this.element = result.element;
    this.complete = result.card.complete;
    if(this.complete) {
      this._stripe.
    }
    this.error = undefined;
  }

  keyUpdated() {
    this._stripe.changeKey(this.stripeKey);
  }

  async payNow() {
    const response = await fetch('/paymentIntent', {
      method: 'POST',
      body: JSON.stringify({amount: 10000, currency: 'usd'}),
      headers: {'content-type': 'application/json'},
    });

    const body = await response.json();
    this._stripe.handleCardPayment(body.client_secret, this.element, { payment_method_data: {
      billing_details: {name: 'Bob Smith'}
    }}).subscribe(result => {
      if(result.error) {
        console.error('got stripe error', result.error);
      } else {
        console.log('payment succeeded');
      }
    });
  }
}
```

Please note you can also manually handle this flow using the paymentIntent methods however that is not the default path and you will have to implement it yourself (see more [here](https://stripe.com/docs/payments/3d-secure))

## Testing
The following command runs unit & integration tests that are in the `tests` folder, and unit tests that are in `src` folder:
```Shell
yarn test
```

## Building
The following command:
```Shell
yarn build
```
- starts _TSLint_ with _Codelyzer_
- starts _AoT compilation_ using _ngc_ compiler
- creates `dist` folder with all the files of distribution

To test the npm package locally, use the following command:
```Shell
yarn publish:dev
```
You can then run the following to install it in an app to test it:
```Shell
yalc link @niteshp/ngx-stripe
```

## Publishing

```Shell
yarn release:patch
or
yarn release:minor
or
yarn release:major
```

## Documentation
To generate the documentation, this starter uses [compodoc](https://github.com/compodoc/compodoc):
```Shell
yarn compodoc
yarn compodoc:serve
```

## License

MIT


CHANGELOG.md
### Changelog

All notable changes to this project will be documented in this file. Dates are displayed in UTC.

Generated by [`auto-changelog`](https://github.com/CookPete/auto-changelog).

#### [v2.0.5](https://github.com/niteshpurohit/ngx-stripe/compare/2.0.4...v2.0.5)

> 10 February 2020

#### [2.0.4](https://github.com/niteshpurohit/ngx-stripe/compare/2.0.3...2.0.4)

> 10 February 2020

- Release 2.0.4 [`1e1853b`](https://github.com/niteshpurohit/ngx-stripe/commit/1e1853bcfc85ccf3326591d2caad44c67a8b2474)

#### [2.0.3](https://github.com/niteshpurohit/ngx-stripe/compare/2.0.2...2.0.3)

> 10 February 2020

- Release 2.0.3 [`ccbcfb6`](https://github.com/niteshpurohit/ngx-stripe/commit/ccbcfb60ab1450091f611c6ac4b1d0ff48462e51)

#### [2.0.2](https://github.com/niteshpurohit/ngx-stripe/compare/2.0.1...2.0.2)

> 10 February 2020

- Release 2.0.2 [`d275897`](https://github.com/niteshpurohit/ngx-stripe/commit/d275897802dfa77aba9a9be81572e6ba9270bb5a)
- change publish config [`e750f76`](https://github.com/niteshpurohit/ngx-stripe/commit/e750f7601d345bd225707abffa94ab7f2d4b2b74)

#### [2.0.1](https://github.com/niteshpurohit/ngx-stripe/compare/2.0.0...2.0.1)

> 10 February 2020

- ng 9 [`a87ceec`](https://github.com/niteshpurohit/ngx-stripe/commit/a87ceec981873e105482118d4fd46a7c744c1231)
- Release 2.0.1 [`6c3fd21`](https://github.com/niteshpurohit/ngx-stripe/commit/6c3fd21999e84aef71fbcc764957f32924e829cd)

#### 2.0.0

> 17 December 2019

- First [`ec61541`](https://github.com/niteshpurohit/ngx-stripe/commit/ec61541975b6c71f59d3ef0bfb8a11a989a68480)
- Release 2.0.0 [`ad52093`](https://github.com/niteshpurohit/ngx-stripe/commit/ad52093cbf04bc988ebf02b3853d0a4dd6802de2)
- Second [`e4ed836`](https://github.com/niteshpurohit/ngx-stripe/commit/e4ed8365a92aceadc176e6deaf1d07b090d461e4)

#### [v1.2.10](https://github.com/niteshpurohit/ngx-stripe/compare/1.2.9...v1.2.10)

> 8 December 2020

- update according to StripeJS [`08a9432`](https://github.com/niteshpurohit/ngx-stripe/commit/08a9432b34cdc0fd1b31d2c77b923b139100bee3)

#### [1.2.9](https://github.com/niteshpurohit/ngx-stripe/compare/1.2.8...1.2.9)

> 8 December 2020

- update according to StripeJS [`9864a20`](https://github.com/niteshpurohit/ngx-stripe/commit/9864a20a422c2f443f1bd633cf8a828f8562c93a)
- Release 1.2.9 [`cb1fde0`](https://github.com/niteshpurohit/ngx-stripe/commit/cb1fde025b9009c088bc1a738cd81a379a7282c1)

#### [1.2.8](https://github.com/niteshpurohit/ngx-stripe/compare/1.2.7...1.2.8)

> 28 October 2020

- build(deps): yarn-audit-fix [`704656d`](https://github.com/niteshpurohit/ngx-stripe/commit/704656d95d35fcc2b80bb82bab3170c0f1db3f33)
- build(material): update material to 10 [`7a85271`](https://github.com/niteshpurohit/ngx-stripe/commit/7a8527114248e73a684e7aa6617527e698c6ccbc)
- Release 1.2.8 [`90ea31b`](https://github.com/niteshpurohit/ngx-stripe/commit/90ea31b98de6969a16faccef75a2b6b771affc2a)

#### [1.2.7](https://github.com/niteshpurohit/ngx-stripe/compare/1.2.6...1.2.7)

> 28 October 2020

- Bump node-fetch from 2.6.0 to 2.6.1 [`#7`](https://github.com/niteshpurohit/ngx-stripe/pull/7)
- build(angular): update angular to 10 [`4a45777`](https://github.com/niteshpurohit/ngx-stripe/commit/4a45777837db5d553c7828085d6940da973e7516)
- build(deps): yarn-audit-fix [`d4bd73c`](https://github.com/niteshpurohit/ngx-stripe/commit/d4bd73c8570f90548643f7c97f70b1eac570136f)
- Create codeql-analysis.yml [`4b69bb4`](https://github.com/niteshpurohit/ngx-stripe/commit/4b69bb45d10e22abffab3453000ea0c501f1059f)

#### [1.2.6](https://github.com/niteshpurohit/ngx-stripe/compare/1.2.5...1.2.6)

> 31 August 2020

- build(deps-dev): bump deps [`e8c059f`](https://github.com/niteshpurohit/ngx-stripe/commit/e8c059f51e8b42c0fa845f22cc5265e1214c965d)
- Release 1.2.6 [`073f060`](https://github.com/niteshpurohit/ngx-stripe/commit/073f0604a20e84a900b7a1038c67bfab5a1aba56)
- docs(fix-import): fix StripeElement import [`4862853`](https://github.com/niteshpurohit/ngx-stripe/commit/48628534aad6b537402e0e9af20779adf78d3883)

#### [1.2.5](https://github.com/niteshpurohit/ngx-stripe/compare/1.2.4...1.2.5)

> 12 August 2020

- Bump npm-registry-fetch from 4.0.3 to 4.0.5 [`#4`](https://github.com/niteshpurohit/ngx-stripe/pull/4)
- Bump elliptic from 6.5.2 to 6.5.3 [`#5`](https://github.com/niteshpurohit/ngx-stripe/pull/5)
- Bump websocket-extensions from 0.1.3 to 0.1.4 [`#3`](https://github.com/niteshpurohit/ngx-stripe/pull/3)
- Release 1.2.5 [`2a05735`](https://github.com/niteshpurohit/ngx-stripe/commit/2a057359f60874e95667857275414124af342ae3)
- Update [`982a9d5`](https://github.com/niteshpurohit/ngx-stripe/commit/982a9d59f0e6544e5962ed971c25856012ce64df)

#### [1.2.4](https://github.com/niteshpurohit/ngx-stripe/compare/1.2.3...1.2.4)

> 17 April 2020

- Bump https-proxy-agent from 2.2.2 to 2.2.4 [`#2`](https://github.com/niteshpurohit/ngx-stripe/pull/2)
- Bump acorn from 5.7.3 to 5.7.4 [`#1`](https://github.com/niteshpurohit/ngx-stripe/pull/1)
- update deps [`19f4576`](https://github.com/niteshpurohit/ngx-stripe/commit/19f4576b30ad395278438b3e19adb046ea4652ee)
- Change scope. [`5962bff`](https://github.com/niteshpurohit/ngx-stripe/commit/5962bffddbc6123ceae43b96b1bb7de3e2ae1ee8)
- Update [`551c47b`](https://github.com/niteshpurohit/ngx-stripe/commit/551c47bc2ef9687593112365cf8e8c07c9dfefe1)

### [1.2.3](https://github.com/niteshpurohit/ngx-stripe/compare/v2.0.5...1.2.3)

> 10 February 2020

- Release 1.2.3 [`3dd91fa`](https://github.com/niteshpurohit/ngx-stripe/commit/3dd91fa4641e16e553fd81c5108e88f7ec6861e9)
- continue v1 [`5ad9233`](https://github.com/niteshpurohit/ngx-stripe/commit/5ad923358e3335acf6fbf78bfc625aa93d682b04)


src/app/app.module.ts
import { NgModule } from '@angular/core';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { MatButtonModule } from '@angular/material/button';
import { MatCardModule } from '@angular/material/card';
import { MatIconModule } from '@angular/material/icon';
import { MatInputModule } from '@angular/material/input';
import { MatSliderModule } from '@angular/material/slider';
import { MatToolbarModule } from '@angular/material/toolbar';
import { BrowserModule } from '@angular/platform-browser';
import { BrowserAnimationsModule } from '@angular/platform-browser/animations';
import { NgxStripeModule } from '@niteshp/ngx-stripe';
import { AppComponent } from './app.component';

@NgModule({
  declarations: [AppComponent],
  imports: [
    BrowserModule,
    BrowserAnimationsModule,
    MatInputModule,
    MatButtonModule,
    MatToolbarModule,
    MatSliderModule,
    MatIconModule,
    MatCardModule,
    FormsModule,
    ReactiveFormsModule,
    NgxStripeModule.forRoot('')
  ],
  providers: [],
  bootstrap: [AppComponent]
})
export class AppModule { }


projects/ngx-stripe/src/lib/index.ts
// Public classes.
export { Element, ElementEventType, ElementOptions, ElementStyleAttributes, ElementType } from './interfaces/element';
export { Elements, ElementsOptions, FontElement } from './interfaces/elements';
export {
  FlowTypes,
  isSourceData,
  Source,
  SourceData,
  SourceParams,
  SourceResult,
  UsageTypes,
} from './interfaces/sources';
export { Options, StripeJS, STRIPE_OPTIONS, STRIPE_PUBLISHABLE_KEY } from './interfaces/stripe';
export * from './interfaces/token';
export { Address, Error } from './interfaces/utils';
export { NgxStripeModule } from './ngx-stripe.module';
export { LazyStripeAPILoader } from './services/api-loader.service';
export { DocumentRef } from './services/document-ref.service';
export { PlatformService } from './services/platform.service';
export { StripeService } from './services/stripe.service';
export { WindowRef } from './services/window-ref.service';
export { StripeCardComponent } from './stripe-card.component';


projects/ngx-stripe/src/lib/interfaces/token.ts
import { Error } from './utils';

export interface Token {
  id: string;
  object: 'token';
  bank_account?: {
    id: string;
    country: string;
    currency: string;
    fingerprint: string;
    object: 'bank_account';
    account_holder_name: string;
    account_holder_type: 'individual' | 'company';
    bank_name: string;
    last4: string;
    routing_number: string;
    status: 'new' | 'validated' | 'verified' | 'verification_failded' | 'errored';
  };
  card?: {
    id: string;
    country: string;
    currency: string;
    fingerprint: string;
    object: 'card';
    address_city: string;
    address_country: string;
    address_line1: string;
    address_line1_check: FieldCheck;
    address_line2: string;
    address_state: string;
    address_zip: string;
    address_zip_check: FieldCheck;
    brand: string;
    cvc_check: FieldCheck;
    dynamic_last4: string;
    exp_month: number;
    exp_year: number;
    funding: 'credit' | 'debit' | 'prepaid' | 'unknown';
    last4: string;
    metadata: { [key: string]: any };
    name: string;
    tokenization_method: 'apple_pay' | 'android_pay';
  };
  client_ip: string;
  livemode: boolean;
  type: 'card' | 'bank_account';
  used: boolean;
}

export type FieldCheck = 'pass' | 'fail' | 'unavailable' | 'unchecked';

export interface CardDataOptions {
  name?: string;
  address_line1?: string;
  address_line2?: string;
  address_city?: string;
  address_state?: string;
  address_zip?: string;
  address_country?: string;
  currency?: string;
}

export interface TokenResult {
  token?: Token;
  error?: Error;
}

export type BankAccount = 'bank_account';

export interface BankAccountData {
  country: string;
  currency: string;
  routing_number: string;
  account_number: string;
  account_holder_name: string;
  account_holder_type: 'individual' | 'company';
}

export type Pii = 'pii';

export interface PiiData {
  personal_id_number: string;
}

export function isBankAccount(account: any): account is BankAccount {
  return account === 'bank_account';
}

export function isBankAccountData(bankAccountData: any): bankAccountData is BankAccountData {
  return (
    'country' in bankAccountData &&
    'currency' in bankAccountData &&
    'routing_number' in bankAccountData &&
    'account_number' in bankAccountData &&
    'account_holder_name' in bankAccountData &&
    'account_holder_type' in bankAccountData &&
    (bankAccountData.account_holder_type === 'individual' || bankAccountData.account_holder_type === 'company')
  );
}

export function isPii(pii: any): pii is Pii {
  return pii === 'pii';
}

export function isPiiData(piiData: any): piiData is PiiData {
  return 'personal_id_number' in piiData;
}

export interface SetupIntent {
  id: string;
  object: 'setup_intent';
  application?: string;
  cancellation_reason?: 'abandoned' | 'requested_by_customer' | 'duplicate';
  client_secret?: string;
  created: number;
  customer?: string;
  description?: string;
  last_setup_error?: Error;
  livemode: boolean;
  metadata: { [key: string]: any };
  next_action: {
    type: 'redirect_to_url' | 'use_stripe_sdk';
    redirect_to_url?: {
      return_url: string;
      url: string;
    };
    use_stripe_sdk: { [key: string]: any };
  };
  on_behalf_of?: string;
  payment_method?: string;
  payment_method_options?: { [key: string]: any };
  status:
    | 'requires_payment_method'
    | 'requires_confirmation'
    | 'requires_action'
    | 'processing'
    | 'canceled'
    | 'succeeded';
  usage: 'on_session' | 'off_session';
}

export interface SetupIntentResult {
  setupIntent?: SetupIntent;
  error?: Error;
}

export interface SetupIntentData {
  payment_method_data?: {
    billing_details?: BillingDetails;
  };
}

export interface BillingDetails {
  address?: Address;
  email?: string;
  name?: string;
  phone?: string;
}

export interface Shipping {
  address: {
    city?: string;
    country?: string;
    line1?: string;
    line2?: string;
    postal_code?: string;
    state?: string;
  };
  name: string;
  carrier?: string;
  phone?: string;
  tracking_number?: string;
}

export interface ConfirmSetupIntentData {
  payment_method_data?: {
    billing_details?: BillingDetails;
    return_url?: string;
  };
}

export interface Address {
  city?: string;
  country?: string;
  line1?: string;
  line2?: string;
  postal_code?: string;
  state?: string;
}

export interface ShippingOptions {
  id: string;
  label: string;
  detail: string;
  amount: number;
}

export interface ShippingAddress {
  country?: string;
  addressLine?: string;
  region?: string;
  city?: string;
  postalCode?: string;
  recipient?: string;
  phone?: string;
  sortingCode?: string;
  dependentLocality?: string;
}

export interface PaymentIntent {
  id: string;
  object: 'payment_intent';
  amount: number;
  amount_capturable: number;
  amount_received: number;
  application?: string;
  application_fee_ammount?: number;
  canceled_at?: number;
  cancellation_reason?: 'abandoned' | 'requested_by_customer' | 'duplicate';
  cancelled_method?: 'automatic' | 'manual';
  capture_method?: 'automatic' | 'manual';
  charges?: Array<{
    object: 'list';
    data: Array<any>;
    has_more: boolean;
    url: string;
  }>;
  client_secret?: string;
  confirmation_method?: 'automatic' | 'manual';
  create: number;
  currency?: string;
  customer?: any;
  description?: string;
  invoice?: string;
  last_payment_error?: any;
  livemode: boolean;
  metadata?: any;
  next_action?: any;
  on_behalf_of?: string;
  payment_method?: string;
  payment_method_options?: any;
  payment_method_types?: Array<string>;
  receipt_email?: string;
  review?: string;
  setup_feture_usage?: string;
  shipping?: Shipping;
  statement_descriptor?: string;
  statement_descriptor_suffix?: string;
  status:
    | 'requires_payment_method'
    | 'requires_confirmation'
    | 'requires_action'
    | 'processing'
    | 'requires_capture'
    | 'canceled'
    | 'succeeded';
  transfer_data?: any;
  transfer_group?: string;
}

export interface PaymentIntentResult {
  paymentIntent?: PaymentIntent;
  error?: Error;
}

export interface PaymentMethodOptions {
  card?: Element;
  billing_detail?: BillingDetails;
  setup_future_usage?: string;
  payment_method_options?: any;
}

export interface HandleCardPaymentOptions {
  payment_method: string | PaymentMethodOptions;
}


projects/ngx-stripe/src/lib/interfaces/stripe.ts
import { InjectionToken } from '@angular/core';
import { Element, RequestElementOptions } from './element';
import { Elements, ElementsOptions } from './elements';
import { SourceData, SourceParams, SourceResult } from './sources';
import {
  BankAccount,
  BankAccountData,
  CardDataOptions,
  ConfirmSetupIntentData,
  Pii,
  PiiData,
  SetupIntentData,
  SetupIntentResult,
  TokenResult,
  HandleCardPaymentOptions,
} from './token';

export const STRIPE_PUBLISHABLE_KEY = new InjectionToken<string>('Stripe Publishable Key');
export const STRIPE_OPTIONS = new InjectionToken<Options>('Stripe Options');

export interface StripeJS {
  elements(options?: ElementsOptions): Elements;
  createToken(el: Element, cardData?: CardDataOptions): Promise<TokenResult>;
  createToken(account: BankAccount, bankAccountData: BankAccountData): Promise<TokenResult>;
  createToken(pii: Pii, piiData: PiiData): Promise<TokenResult>;
  createSource(el: Element, sourceData?: SourceData): Promise<SourceResult>;
  createSource(sourceData: SourceData): Promise<SourceResult>;
  paymentRequest(options: RequestElementOptions);
  handleCardSetup(clientSecret: string, el: Element, cardSetupOptions?: SetupIntentData): Promise<SetupIntentResult>;
  handleCardAction(clientSecret: string);
  handleCardPayment(clientSecret: string, data: HandleCardPaymentOptions, options: any);
  retrievePaymentIntent(clientSecret: string);
  confirmSetupIntent(clientSecret: string, el: Element, data: ConfirmSetupIntentData);
  retrieveSetupIntent(clientSecret: string);
  retrieveSource(source: SourceParams): Promise<SourceResult>;
}

export interface Options {
  stripeAccount?: string;
}


src/app/app.component.ts
import { AfterContentInit, Component } from '@angular/core';
import { FormControl, Validators } from '@angular/forms';
import {
  Element as StripeElement,
  ElementOptions,
  ElementsOptions,
  StripeService
} from '@niteshp/ngx-stripe';
import { PaymentRequestButtonStyle, RequestElementOptions } from 'projects/ngx-stripe/src/lib/interfaces/element';
import { environment } from 'src/environments/environment';

@Component({
  selector: 'app-root',
  templateUrl: './app.component.html',
  styleUrls: ['./app.component.css']
})
export class AppComponent implements AfterContentInit {
  error: any;
  stripeKey = new FormControl('', [Validators.required, validateStripeTestKey]);
  keyError = 'Must be a valid publishable test key (e.g. beings with pk_test_)';
  complete = false;
  complete2 = false;
  element: StripeElement;
  cardOptions: ElementOptions = {
    style: {
      base: {
        iconColor: '#276fd3',
        color: '#31325F',
        lineHeight: '40px',
        fontWeight: 300,
        fontFamily: '"Helvetica Neue", Helvetica, sans-serif',
        fontSize: '18px',
        '::placeholder': {
          color: '#CFD7E0'
        }
      }
    }
  };

  requestOptions: RequestElementOptions = {
    country: 'US',
    currency: 'usd',
    requestPayerName: true,
    requestPayerEmail: true,
    requestPayerPhone: true,
    requestShipping: true,
    total: {
      amount: 10000,
      label: 'Donate to the things'
    }
  };

  styles: PaymentRequestButtonStyle = {
    type: 'donate',
    theme: 'light',
    height: '64px'
  };

  elementsOptions: ElementsOptions = {
    locale: 'en'
  };

  public pay = 100;

  token: any = {};

  constructor(private _stripe: StripeService) {
    this.stripeKey.setValue('...');
  }

  ngAfterContentInit() {
    this.keyUpdated();
  }

  cardUpdated(result) {
    console.log(result);
    this.element = result.element;
    this.complete = result.complete;
    this.error = undefined;
  }

  cardUpdated2(result) {
    console.log('update second card', result);
    this.element = result.element;
    this.complete2 = result.complete;
    this.error = undefined;
  }

  keyUpdated() {
    if (this.stripeKey.valid && this.stripeKey.value.length > 0) {
      this._stripe.changeKey(this.stripeKey.value);
    } else if (environment.stripeKey) {
      this._stripe.changeKey(environment.stripeKey);
    }
  }

  updatePay(amount: number) {
    this.pay = amount;
    this.requestOptions = {
      ...this.requestOptions,
      total: {
        ...this.requestOptions.total,
        amount: amount * 100
      }
    };
  }

  requestUpdated(result) {
    console.log('updated token ', result);
    this.token = result;
    // if the server completes payment call this
    result.complete('success');
    // if the server fails to complete payment call this instead
    // result.complete('fail');
  }

  updateShippingAddress(result) {
    console.log('shipping address change', result);
    result.updateWith({
      status: 'success',
      shippingOptions: [
        {
          id: 'free-shipping',
          label: 'Free shipping',
          detail: 'Arrives in 5 to 7 days',
          amount: 0
        },
        {
          id: 'express',
          label: 'Express shipping',
          detail: 'Arrives in 1 to 2 days',
          amount: 1000
        }
      ]
    });
  }
  updateShippingOption(result) {
    console.log('shipping option change', result);
    result.updateWith({
      status: 'success'
    });
  }

  getCardToken() {
    this._stripe
      .createToken(this.element, {
        name: 'tested_ca',
        address_line1: '123 A Place',
        address_line2: 'Suite 100',
        address_city: 'Irving',
        address_state: 'BC',
        address_zip: 'VOE 1H0',
        address_country: 'CA'
      })
      .subscribe(result => {
        this.token = result;
      });
  }
}

function validateStripeTestKey(control: FormControl) {
  if (control.value && control.value.match(/^pk_test_.+/)) {
    return;
  }
  return { valid: false };
}
